File: parser.cpp
#include <iostream>
#include <vector>
#include <stdlib.h>
#include "parser.hpp"

// Parses the lines and creates a corresponding matrix of Fields.
std::vector<std::vector<Field>> Parser::parse(std::vector<std::string> lines) {
    std::cout << std::endl << "Parsing input:" << std::endl;
    for (auto line : lines) {
        std::cout << line << std::endl;
    }

    // Determine number of rows and columns
    auto rows = lines.size();
    if (rows == 0) {
        std::cerr << "No rows in screen!" << std::endl;
        abort();
    }
    auto cols = max_line_length(lines);
    if (cols == 0) {
        std::cerr << "No columns in screen!" << std::endl;
        abort();
    }

    return parse_screen(lines, rows, cols);
}

// Returns the length of the longest string in given vector of strings.
size_t Parser::max_line_length(const std::vector<std::string> &lines) {
    size_t max = 0;
    for (const auto &line : lines) {
        if (line.length() > max) {
            max = line.length();
        }
    }
    return max;
}

// Parses the lines and creates a corresponding <rows> x <cols> matrix of Fields. Removes the outmost columns and rows, since they are walls.
std::vector<std::vector<Field>> Parser::parse_screen(const std::vector<std::string> lines, const size_t rows, const size_t cols) {

    std::vector<std::vector<Field>> screen = create_matrix(rows-2, cols-2, Field::EMPTY);

    for (size_t i = 0; i < rows-2; i++) {
        std::string s = lines[i+1];
        std::vector<Field> &row = screen[i];

        for (size_t j = 0; j < cols-2 && j < s.length(); j++) {
            row[j] = field_functions::convertFromScreenInput(s[j+1]);
        }
    }
    return screen;
}

// create a 'matrix' (vector of vectors) of size <rows> x <cols>, filled with <initial_value>
std::vector<std::vector<Field>> Parser::create_matrix(const size_t rows, const size_t cols, const Field initial_value) {
    std::vector<std::vector<Field>> result;
    for (size_t i = 0; i < rows; i++) {
        std::vector<Field> row;
        row.assign(cols, initial_value);
        result.push_back(row);
    }
    return result;
}

File: src/main/java/sokoban/Parser.java
package sokoban;

public class Parser {

  public Field[][] parse(final String[] lines) {
    System.out.println("Parsing input:");
    for (String line : lines) {
      System.out.println(line);
    }

    // Determine number of rows and columns
    int rows = lines.length;
    if (rows == 0) {
      throw new IllegalArgumentException("No rows in screen!");
    }

    int cols = maxLineLength(lines);
    if (cols == 0) {
      throw new IllegalArgumentException("No columns in screen!");
    }

    return parseScreen(lines, rows, cols);
  }

  private int maxLineLength(final String[] lines) {
    int max = 0;
    for (String line : lines) {
      max = Math.max(max, line.length());
    }
    return max;
  }

  private Field[][] parseScreen(final String[] lines, final int rows, final int cols) {
    Field[][] screen = createMatrix(rows - 2, cols - 2, Field.EMPTY);

    for (int i = 0; i < rows - 2; i++) {
      String s = lines[i + 1];
      Field[] row = screen[i];

      for (int j = 0; j < cols - 2 && j < s.length(); j++) {
        row[j] = Field.convertFromScreenInput(s.charAt(j + 1));
      }
    }

    return screen;
  }

  private Field[][] createMatrix(final int rows, final int cols, final Field initialValue) {
    Field[][] result = new Field[rows][cols];
    for (int i = 0; i < result.length; i++) {
      for (int j = 0; j < result[i].length; j++) {
        result[i][j] = initialValue;
      }
    }
    return result;
  }

}
